### PR TITLE
Replace `flake8`, `isort`, `pylint` and `black` by `ruff`

### LINKED ISSUE #1670: Introducing ruff as a linter to replace pylint, flake8, isort and black.
## Description
 
Kedro recently made the switch from pylint, flake8, isort, and black to ruff which is a lot faster 

Below PRs describe how it's done on Kedro Framework
[PR 2634](https://github.com/kedro-org/kedro/pull/2634)
[PR 3435](https://github.com/kedro-org/kedro/pull/3435)

This ticket is to the same on Kedro-viz backend

## Checklist

- [x] Include labels so that we can categorise your feature request

### PR DESCRIPTION
**Pull request by yury-fedotov**
## Description

Closes #1670 

## Development notes

This PR brings `ruff` to `kedro-viz` to replace:

- `flake8`
- `isort`
- `pylint`
- `black`

It introduces a central `ruff.toml` file in repo root, to act as a single source of truth for Python codestyle management, and leverages `ruff`'s hierarchical configuration feature to make this config applicable to both the `demo-project` and `package` parts of the codebase.

> [!NOTE]  
> Another enhancement is that linter/formatter configuration, such as:
> - Directories in scope
> - Active and ignored rulesets
>
> Is now defined **in a single file, once**, as opposed to being duplicated in `Makefile`, `pre-commit` config etc., which should streamline further code quality management by simplifying addition of rulesets, or directories in scope.

## Note for reviewers

While the diff from this PR is quite big (72 files), most of that is very simple auto-fixes from `ruff format` that are slightly different from `black`. So reviewing should not be very hard. 

## Testing plan

- [x] `make lint` passes locally
- [x] `make pytest` passes locally
- [x] Waiting for maintainers to release `merge-gatekeeper` to run full CI
- [x] CI to pass
- [x] Maintainers to test locally

## Other tasks

- [x] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file

### CODE REVIEW

File: package/tests/conftest.py [hunk 1]
```diff
@@ -420,10 +420,10 @@ def example_catalog():
 @pytest.fixture
 def example_transcoded_pipelines():
     def process_data(raw_data, train_test_split):
-        ...
+        pass
```
Comment (yury-fedotov) [inline package/tests/conftest.py:423]: Otherwise, the ruff formatter reformats that to one liner:

```python
def process_data(raw_data, train_test_split): ...
```

And `test_task_node_metadata` in `package/tests/test_api/test_rest/test_responses.py` fails. See my modification to that test, to make it compatible with `pass`, below.

File: Makefile
Comment (yury-fedotov) [inline Makefile:1]: Note how this simplifies after we define all settings, including directories and rulesets, **once, in a central `ruff.toml` configuration file**. This allows all "parties" that use `ruff`:

- Developers
- CI
- Pre-commit hooks

To run just `ruff check` and `ruff format` from anywhere in the repo, and it applies a corresponding configuration automatically.

File: demo-project/.pre-commit-config.yaml [hunk 2]
```diff
@@ -14,31 +14,11 @@ repos:
           - id: check-json # Checks json files for parseable syntax.
           - id: check-case-conflict # Check for files that would conflict in case-insensitive filesystems
           - id: check-merge-conflict # Check for files that contain merge conflict strings.
-          - id: debug-statements # Check for debugger imports and py37+ `breakpoint()` calls in python source.
```
Comment (yury-fedotov) [inline demo-project/.pre-commit-config.yaml:17]: `ruff`'s `T100` ruleset, which I'm enabling, checks and autofixes that, so this pre-commit hook becomes redundant.

File: ruff.toml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+"package/kedro_viz/api/graphql/types.py" = ["D"]
+"package/kedro_viz/data_access/repositories/graph.py" = ["D"]
+"package/kedro_viz/data_access/repositories/runs.py" = ["D"]
+"demo-project/*.py" = ["D", "ARG", "PLW0603"] # Allow unused arguments in node functions for them to generate constant outputs, but mimic the DAG and technically have inputs.
+
+[lint.mccabe]
+max-complexity = 18
+
+[lint.pylint]
+max-args = 12
```
Comment (yury-fedotov) [inline ruff.toml:69]: I checked all arguments that your `pylint`, `isort` and `flake8` configs used, and they are all default except for those

File: demo-project/src/demo_project/__init__.py
Comment (yury-fedotov) [inline demo-project/src/demo_project/__init__.py:1]: Diffs like this, or e.g. adding a blank line between module docstring and first import line, is what `ruff format` does automatically due to its minor differences with `black`. There's no config for `ruff format` to make it fully backwards compatible.

File: demo-project/src/demo_project/pipelines/data_ingestion/pipeline.py [hunk 4]
```diff
@@ -27,21 +27,21 @@ def create_pipeline(**kwargs) -> Pipeline:
...
+                tags="companies",
             ),
             node(
                 func=apply_types_to_shuttles,
                 inputs="shuttles",
                 outputs="int_typed_shuttles@pandas1",
-                name='apply_types_to_shuttles',
-                tags='shuttles'
+                name="apply_types_to_shuttles",
+                tags="shuttles",
```
Comment (yury-fedotov) [inline demo-project/src/demo_project/pipelines/data_ingestion/pipeline.py:38]: Another autofix from `ruff format` - makes sense, since now single and double quotes are mixes within the same `node` definition. I'm curious why previous formatter didn't flag that.

File: demo-project/src/demo_project/pipelines/reporting/pipeline.py [hunk 5]
```diff
@@ -8,10 +8,10 @@
 from demo_project.pipelines.reporting.nodes import (
     create_feature_importance_plot,
     create_matplotlib_chart,
+    get_top_shuttles_data,
     make_cancel_policy_bar_chart,
     make_price_analysis_image,
     make_price_histogram,
-    get_top_shuttles_data,
```
Comment (yury-fedotov) [inline demo-project/src/demo_project/pipelines/reporting/pipeline.py:14]: I'm curious why `isort` didn't catch that in the past...

File: package/kedro_viz/api/apps.py [hunk 6]
```diff
@@ -42,7 +43,7 @@ def _create_base_api_app() -> FastAPI:
         @app.middleware("http")
         async def set_secure_headers(request, call_next):
             response = await call_next(request)
-            secure_headers.framework.fastapi(response)  # pylint: disable=no-member
```
Comment (yury-fedotov) [inline package/kedro_viz/api/apps.py:45]: I have deleted all `pylint: disable...` directives from the entire codebase.

File: package/ruff.toml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+extend = "../ruff.toml"
+
+[lint.isort]
+known-first-party = ["kedro_viz"]
```
Comment (rashidakanchwala) [inline package/ruff.toml:4]: should we also add kedro here as third party?

File: demo-project/ruff.toml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+extend = "../ruff.toml"
+
+[lint.isort]
```
Comment (ravi-kumar-pilla) [inline demo-project/ruff.toml:3]: Is this a best practice to have ruff.toml ? Can we do this in demo-project/pyproject.toml -> table [tool.ruff.lint.isort] ?

File: demo-project/src/demo_project/pipelines/data_ingestion/pipeline.py (hunk 4 above)
Comment (ravi-kumar-pilla) [reply demo-project/src/demo_project/pipelines/data_ingestion/pipeline.py:38]: The formatter was [run](https://github.com/kedro-org/kedro-viz/pull/2149/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L30) only on `package/` . The `demo-project` is outside of the formatter.

**Review by rashidakanchwala — APPROVED** [review]
LGTM. thanks a lott @yury-fedotov

File: demo-project/ruff.toml (hunk 8 above)
Comment (yury-fedotov) [reply demo-project/ruff.toml:3]: `ruff` supports both `pyproject.toml` section and dedicated `ruff.toml` - see [here](https://docs.astral.sh/ruff/configuration/#config-file-discovery). So yes what you propose is feasible.

There is no explicit recommendation in `ruff` docs on which to prefer, and I have not heard of what's generally considered best practice. My thinking in general is that whenever possible, I'd prefer a dedicated config for every tool, because it:

1. Makes each config file have isolated responsibility
2. Simplifies version control a bit - e.g. 2 PRs where one changes dependencies and another changes the lint rulesets wouldn't touch the same file
3. Makes `pyproject.toml` smaller and more manageable

So I use `ruff.toml` in all of my projects, and never put that in `pyproject.toml` because of above.

Let me know if you disagree & still prefer the `pyproject.toml` - I can change to that. If so, then probably we'd need to configure `ruff` via `pyproject.toml` not only in `demo-project` but also in repo root and `package`.

Comment (yury-fedotov) [reply demo-project/ruff.toml:3]: I came up with a better argument :)

1. It's useful to have `ruff` configuration **in repo root**, so that both `package`, `demo-project` or other **packages** can inherit it
2. Repo root is not a Python package, so `ruff.toml` is preferred over `pyproject.toml` in root
3. (1 + 2) sort of require us to use `ruff.toml` in the root, so for consistency, I suggest using `ruff.toml` files in `package` and `demo-project` too.

Comment (ravi-kumar-pilla) [reply demo-project/ruff.toml:3]: Sure, I am happy to have ruff.toml, I wanted to understand if this was a better practice suggested from ruff. Thanks for the PR

**Review by ravi-kumar-pilla — APPROVED** [review]
Thank you @yury-fedotov 💯

File: ruff.toml [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    "F401",
+    "B030", # Except handler is something other than exception class
+    "C405", # Inconsistent definition of literal collections
+    "PLR2004", # Magic values in comparisons
+]
+
+[lint.per-file-ignores]
+"*/cli_steps.py" = ["B011"] # assert False instead of AssertionError
+"*/base_deployer.py" = ["B024"] # ABCs without abstract methods
+"package/kedro_viz/__init__.py" = ["B028"] # Risky usage of positional arguments
```
Comment (ankatiyar) [inline ruff.toml:41]: Q: does it make sense to ignore some of these specific issues in the code with noqa inline instead of here? Can keep the ignores that apply to multiple files here.

**Review by ankatiyar — APPROVED** [review]
Overall looks good but left a question regarding the ignoring of errors. 
Thanks @yury-fedotov! 💯

File: ruff.toml (hunk 9 above)
Comment (rashidakanchwala) [reply ruff.toml:41]: I definitely wouldn't want to ignore the 'broad exception' rule in general, because I think we should avoid it as much as possible. I guess some others like # noqa: ARG002 should be fine.

Comment (yury-fedotov) [reply ruff.toml:41]: @rashidakanchwala 

_Broad exception_ rule has `BLE` code in `ruff` and is selected for entire repo ✅  (See line 14 of this file)

Comment (yury-fedotov) [reply ruff.toml:41]: @ankatiyar 

Good question!

The way I implemented it is as follows:

1. Whatever has an inline NOQA before that's applied to 1-2 places, stays as inline NOQA.
2. Whatever was a file-level `pylint: disable` statement at the beginning of the file is moved to `ruff.toml` from that separate file.

Let me explain the rationale for two.

Consider two implementations, using `package/kedro_viz/integrations/kedro/sqlite_store.py` as an example:

**Before**

```python
# pylint: broad-exception-caught
```

At the beginning of that particular file.

**After**

```toml
"package/kedro_viz/integrations/kedro/sqlite_store.py" = ["BLE"]
```

In `ruff.toml`

----

- In both cases, we essentially store the same directive - disable a particular rule in a particular file.
- However (2) makes it more explicit and centralizes all those directives, which would otherwise be scattered across tens of files, in a **single file**.
- Long term, that helps gradually make the codebase adhere to those rulesets, and delete those `per-file-ignores`. As opposed to distributed inline NOQAs that typically just never addressed.

**Comment by rashidakanchwala** [comment]
@yury-fedotov , feel free to merge this PR :)

**Comment by yury-fedotov** [comment]
> @yury-fedotov , feel free to merge this PR :)

I don't have write access

**Comment by rashidakanchwala** [comment]
no worries, let me merge. Once again Thank you!, this is big improvement to our dev workflow!
